### ISSUE TITLE
`self.device` is not correct during `setup`

### ISSUE THREAD

**Issue by carmocca** [PyTorchLightning/pytorch-lightning#13108]
### To Reproduce

```python
import os

from pytorch_lightning import Trainer
from pytorch_lightning.demos.boring_classes import BoringModel


class BugReportModel(BoringModel):
    def setup(self, stage=None):
        # current workaround
        assert self.trainer.strategy.root_device.type == "cuda"
        # error! it is on cpu
        assert self.device.type == "cuda"


def run():
    model = BugReportModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        fast_dev_run=True,
        accelerator="gpu",
        devices=1,
    )
    trainer.fit(model)


if __name__ == "__main__":
    run()
```

### Expected behavior

`lightning_module.device` and `strategy.device` point to the same device

### Additional context

Reported in Slack by Joshua Chia

**Comment by awaelchli** [comment]
We cannot expect this to be true at any point in time:

> lightning_module.device and strategy.device point to the same device

The value of `strategy.device` is available as soon as the trainer instantiation finishes. Between then and `trainer.fit`, it is clear that the `lightning_module.device` cannot possibly have the updated value already. However, we can expect that after a certain point in the hook call timeline, these two become identical. When is that point?

We know that it certainly must be after `configure_sharded_model()`, because only then do we know that all layers are created completely. `configure_sharded_model` runs after the `setup()` hook, so that excluses `setup()` from being able to meet the requirement. In essence, the sequence is this:

```
setup()
configure_sharded_model()
move_model_to_device()
```

Could one just change the order to

```
configure_sharded_model()
move_model_to_device()
setup()
```
??

I'm not so sure if a breaking change like this is worth fixing this "bug".

This argumentation assumes that the value of `lightning_module.device` represents the true current placement of the module layers. One could in theory set `lightning_module.device` equal to `strategy.device` without actually moving the module layers, just for the sake of having them equal early on in the setup phase of fit, but that would be misleading IMO.

**Comment by Daniel451** [comment]
Is there any update on this?

@carmocca what type of "hotfix" are you using so far? I tried it with the good old `lit_model = lit_model.to(device)` so far but this is quite counter-intuitive.

@awaelchli having `setup()` and the end would obviously fix this problem but isn't it the case that dataloaders need distribution information when the sharding happens? So if you would configure something w.r.t. dataloaders in `setup()` this would have to happen before `configure_sharded_model()`. Might be totally wrong tho, maybe @Borda has an idea?

**Comment by carmocca** [comment]
> Could one just change the order to

We can't because `setup` can define the model or part of it.

> just for the sake of having them equal early on in the setup phase of fit, but that would be misleading IMO

I get that it might not be the true placement, especially under sharding settings, but I think it's a better solution for the general cases and would avoid errors under them.

@Daniel451 The workaround is using `self.trainer.strategy.root_device` instead.

**Comment by awaelchli** [comment]
@carmocca Do you just want to

```py
for module in pl_module.modules():
    if isinstance(module, LightningModule):
        module._device = trainer.strategy.root_device
```
(pseudocode). You are saying this is something we should do before all the hooks before setup()?

**Comment by carmocca** [comment]
Yes. Do you see any issues with that other than the shortcomings you described in https://github.com/Lightning-AI/lightning/issues/13108#issuecomment-1364755197?

**Comment by awaelchli** [comment]
I think we should try it and see if any significant tests fail. If not, I'm good with this solution.

**Comment by maxfreu** [comment]
I had the problem that parts of my model weren't moved to the device. First I tried to move it in setup(), but now my workaround is to move them manually in on_fit_start, as the device ist correctly initialized there.
